prog/dftb+/lib_dftb/rangeseparated.F90

```diff
@@ -693,16 +699,181 @@ subroutine updateHamiltonianBlock(descM, descN, pSma, pSbN, pPab)
+    allocate(Hlr(nAOs,nAOs))
+
+    call allocateAndInit(this, iSquare, orb, overlap, densSqr, HH, Smat, Dmat, LRgammaAO)
+    call evaluateHamiltonian(this, Smat, Dmat, LRgammaAO, Hlr)
+    HH(:,:) = HH(:,:) + Hlr(:,:)
+    this%lrenergy = this%lrenergy + 0.5_dp * sum(Dmat(:,:)*Hlr(:,:))
+
+  contains
+
+    subroutine allocateAndInit(this, iSquare, orb, overlap, densSqr, HH, Smat, Dmat, LRgammaAO)
+
+      !> class instance
+      type(RangeSepFunc), intent(inout) :: this
+
+      !> Position of each atom in the rows/columns of the square matrices. Shape: (nAtom)
+      integer, dimension(:), intent(in) :: iSquare
+
+      !> Orbital information.
+      type(TOrbitals), intent(in) :: orb
+
+      real(dp), intent(in) :: overlap(:,:)
+      real(dp), intent(in) :: densSqr(:,:)
+
+      real(dp), intent(inout) :: HH(:,:)
+
+      real(dp), intent(out) :: Smat(:,:)
+      real(dp), intent(out) :: Dmat(:,:)
+      real(dp), intent(out) :: LRgammaAO(:,:)
+
+      integer, allocatable :: get_iat(:)
+
+      integer :: nAtom, nAOs
+      integer :: iAt, G1, G2, mu, nu
+
+      nAtom = size(orb%nOrbAtom,dim=1)
+      nAOs = size(overlap,dim=1)
+
+      allocate(get_iat(nAOs))
+
+      ! Symmetrize Hamiltonian, overlap, density matrices
+      call symmetrizeSquareMatrix(HH)
+      Smat(:,:) = overlap(:,:)
+      call symmetrizeSquareMatrix(Smat)
+      Dmat(:,:) = densSqr(:,:)
+      call symmetrizeSquareMatrix(Dmat)
+
+      do mu = 1, nAOs
+        do iAt = 1, nAtom
+          if (mu > iSquare(iAt)-1 .and. mu <= iSquare(iAt+1)-1) then
+            get_iat(mu) = iAt
+          end if
+        end do
+      end do
+
+      ! Get long-range gamma variable
+      LRgammaAO(:,:) = 0.0_dp
+      do mu = 1, nAOs
+        do nu = mu, nAOs
+          ! Find proper atom index
+          G1 = get_iat(mu)
+          G2 = get_iat(nu)
+          LRgammaAO(nu,mu) = this%lrGammaEval(G2,G1)
+          if (nu /= mu) LRgammaAO(mu,nu) = LRgammaAO(nu,mu)
+        end do
+      end do
+
+    end subroutine allocateAndInit
+
+    subroutine evaluateHamiltonian(this, Smat, Dmat, LRgammaAO, Hlr)
+
+      !> class instance
+      type(RangeSepFunc), intent(inout) :: this
+
+      real(dp), intent(in) :: Smat(:,:)
+      real(dp), intent(in) :: Dmat(:,:)
+      real(dp), intent(in) :: LRgammaAO(:,:)
+
+      real(dp), intent(out) :: Hlr(:,:)
+
+      real(dp), allocatable :: Hmat(:,:)
+      real(dp), allocatable :: tmp_mat(:,:)
+
+      integer :: nAOs, mu, nu
+
+      nAOs = size(overlap,dim=1)
+
+      allocate(Hmat(nAOs,nAOs))
+      allocate(tmp_mat(nAOs,nAOs))
+
+      Hlr(:,:) = 0.0_dp
+
+      tmp_mat(:,:) = 0.0_dp
+      call gemm(tmp_mat,Smat,Dmat,1.0_dp,0.0_dp,'N','N')
+      Hmat(:,:) = 0.0_dp
+      call gemm(Hmat,tmp_mat,Smat,1.0_dp,0.0_dp,'N','N')
+      do mu = 1, nAOs
+        do nu = mu, nAOs
+          Hmat(mu,nu) = Hmat(mu,nu) * LRgammaAO(mu,nu)
+          if (mu /= nu) Hmat(nu,mu) = Hmat(mu,nu)
+        end do
+      end do
```

Comment:
Why not replacing the loop with `Hmat(:,:) = Hmat * LRgammaAO`? Costs a few multiplications more, but we can spare the `if`. You could do it at all 4 occasions below.